local/getDevice.go

```diff
@@ -4,31 +4,117 @@ import (
 	"context"
 	"fmt"
 	"sync"
+	"time"
 
-	"github.com/go-ocf/sdk/local/device"
-	"github.com/go-ocf/sdk/local/resource"
+	gocoap "github.com/go-ocf/go-coap"
+	"github.com/go-ocf/sdk/schema"
 )
 
+// GetDevice performs a multicast and returns a device object if the device responds.
+func (c *Client) GetDevice(ctx context.Context, deviceID string) (*Device, schema.ResourceLinks, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	multicastConn := DialDiscoveryAddresses(ctx, c.errFunc)
+	defer func() {
+		for _, conn := range multicastConn {
+			conn.Close()
+		}
+	}()
+
+	h := newDeviceHandler(deviceID, c.tlsConfig, c.retryFunc, c.retrieveTimeout, c.errFunc, c.resolveEndpointsFunc, cancel)
+	err := DiscoverDevices(ctx, multicastConn, h)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not get the device %s: %v", deviceID, err)
+	}
+	d, dlinks := h.Device()
+	if d == nil {
+		return nil, nil, fmt.Errorf("no response from the device %s", deviceID)
+	}
+	return d, dlinks, nil
+}
+
+func newDeviceHandler(
+	deviceID string,
+	tlsConfig *TLSConfig,
+	retryFunc RetryFunc,
+	retrieveTimeout time.Duration,
+	errFunc ErrFunc,
+	resolveEndpointsFunc ResolveEndpointsFunc,
+	cancel context.CancelFunc,
+) *deviceHandler {
+	return &deviceHandler{
+		deviceID:             deviceID,
+		tlsConfig:            tlsConfig,
+		retryFunc:            retryFunc,
+		retrieveTimeout:      retrieveTimeout,
+		errFunc:              errFunc,
+		resolveEndpointsFunc: resolveEndpointsFunc,
+		cancel:               cancel,
+	}
+}
+
 type deviceHandler struct {
-	deviceID string
-	cancel   context.CancelFunc
+	deviceID             string
+	tlsConfig            *TLSConfig
+	retryFunc            RetryFunc
+	retrieveTimeout      time.Duration
+	errFunc              ErrFunc
+	resolveEndpointsFunc ResolveEndpointsFunc
+	cancel               context.CancelFunc
 
-	client *device.Client
-	lock   sync.Mutex
-	err    error
+	lock        sync.Mutex
+	device      *Device
+	deviceLinks schema.ResourceLinks
+	err         error
 }
 
-func newDeviceHandler(deviceID string, cancel context.CancelFunc) *deviceHandler {
-	return &deviceHandler{deviceID: deviceID, cancel: cancel}
+func (h *deviceHandler) Device() (*Device, schema.ResourceLinks) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+	return h.device, h.deviceLinks
 }
 
-func (h *deviceHandler) Handle(ctx context.Context, client *device.Client) {
+func (h *deviceHandler) Handle(ctx context.Context, conn *gocoap.ClientConn, links schema.ResourceLinks) {
+	conn.Close()
 	h.lock.Lock()
 	defer h.lock.Unlock()
-	if client.DeviceID() == h.deviceID {
-		h.client = client
-		h.cancel()
+
+	link, ok := links.GetResourceLink("/oic/d")
+	if !ok {
+		h.err = fmt.Errorf("cannot get link to /oic/d")
+		return
+	}
+	deviceID := link.GetDeviceID()
+	if deviceID == "" {
+		h.err = fmt.Errorf("cannot determine deviceID")
+		return
 	}
+
+	if h.device != nil || deviceID != h.deviceID {
+		return
```

Comment:
Isn't `deviceID != h.deviceID` is an invalid state that should be reported as an error?

Reply:
no because multicast